qualtran/bloqs/rotations/programmable_ancilla_rotation.py

```diff
@@ -0,0 +1,173 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from collections import Counter
+from functools import cached_property
+
+import numpy as np
+import sympy
+from attrs import field, frozen
+
+from qualtran import Bloq, bloq_example, BloqBuilder, QBit, Register, Side, Signature, SoquetT
+from qualtran.bloqs.basic_gates import CNOT, Hadamard, Rz, XGate, ZPowGate
+from qualtran.bloqs.basic_gates._shims import Measure
+from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+from qualtran.symbolics import ceil, is_symbolic, log2, SymbolicFloat, SymbolicInt
+
+
+@frozen
+class ZPowProgrammedAncilla(Bloq):
+    r"""Resource qubit with state $\frac1{\sqrt2} (|0\rangle + e^{i \pi t} |1\rangle)$.
+
+    Args:
+        exponent: value of $t$.
+        eps: precision of the synthesized state.
+    """
+    exponent: SymbolicFloat
+    eps: SymbolicFloat = 1e-11
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature([Register("q", QBit(), side=Side.RIGHT)])
+
+    def build_composite_bloq(self, bb: 'BloqBuilder') -> dict[str, 'SoquetT']:
+        q = bb.allocate(dtype=QBit())
+        q = bb.add(Hadamard(), q=q)
+        q = bb.add(ZPowGate(self.exponent, eps=self.eps), q=q)
+        return {'q': q}
+
+
+@bloq_example
+def _zpow_programmed_ancilla() -> ZPowProgrammedAncilla:
+    zpow_programmed_ancilla = ZPowProgrammedAncilla(np.pi / 4)
+    return zpow_programmed_ancilla
+
+
+@bloq_example
+def _zpow_programmed_ancilla_symb() -> ZPowProgrammedAncilla:
+    t = sympy.Symbol(r"t")
+    zpow_programmed_ancilla_symb = ZPowProgrammedAncilla(t)
+    return zpow_programmed_ancilla_symb
+
+
+@frozen
+class ZPowUsingProgrammedAncilla(Bloq):
+    r"""Single qubit ZPow rotation using resource states.
+
+    This bloq applies a single qubit Z**t rotation only using ZPow resource states and
+    clifford gates. It is designed to exit early as soon as a measurement succeeds.
```

Comment:
Please add a brief explanation of the strategy used to apply `Z**t` and the corresponding costs.